http2/5_1_1_stream_identifiers.go

```diff
@@ -68,5 +68,64 @@ func StreamIdentifiers() *spec.TestGroup {
 		},
 	})
 
+    // An endpoint that receives an unexpected stream identifier
+    // MUST respond with a connection error (Section 5.4.1) of
+    // type PROTOCOL_ERROR.
+    tg.AddTestCase(&spec.TestCase{
+        Desc:        "Sends stream identifier that takes more than 31 bits",
+        Requirement: "The endpoint MUST response with a connection error of type PROTOCOL_ERROR.",
+        Run: func(c *config.Config, conn *spec.Conn) error {
+            err := conn.Handshake()
+            if err != nil {
+                return err
+            }
+
+            headers := spec.CommonHeaders(c)
+            hp := http2.HeadersFrameParam{
+                StreamID:      0x80000000, // 32 bits long value
```

Comment:
It is impossible to send 32 bit long stream identifier, because its frame field is only 31 bit long.
Value 0x80000000 sets frame fields to:
```
      R = 1
      Stream Identifier = 0
```

And value 0x80000005 will make valid frame with Stream Identifier = 5.

Reply:
No, it doesn't work that way.
With:
```diff
diff --git a/http2/5_1_1_stream_identifiers.go b/http2/5_1_1_stream_identifiers.go
index f7eae75..c40f5d4 100644
--- a/http2/5_1_1_stream_identifiers.go
+++ b/http2/5_1_1_stream_identifiers.go
@@ -4,6 +4,7 @@ import (
        "github.com/summerwind/h2spec/config"
        "github.com/summerwind/h2spec/spec"
        "golang.org/x/net/http2"
+       "fmt"
 )

 func StreamIdentifiers() *spec.TestGroup {
@@ -87,6 +88,7 @@ func StreamIdentifiers() *spec.TestGroup {
                 EndHeaders:    true,
                 BlockFragment: conn.EncodeHeaders(headers),
             }
+            fmt.Printf("%+v\n", hp)
             conn.WriteHeaders(hp)

             return spec.VerifyConnectionError(conn, http2.ErrCodeProtocol)

```
the final structure appears to be `{StreamID:2147483648 BlockFragment:[130 135 132 65 133 2 224 92 11 131] EndStream:true EndHeaders:true PadLength:0 Priority:{StreamDep:0 Exclusive:false Weight:0}}
`, so it actually tries to assign this stream identifier. It causes `http2.ErrCodeProtocol` and it gets verified further on as expected.

Reply:
`PROTOCOL_ERROR` response is caused by `StreamID = 0` value.
There is already test for that case: https://github.com/tempesta-tech/h2spec/blob/53a54456df85e2385b0ba7237ae21c44414d6d1b/http2/6_2_headers.go#L99

Setting Stream ID to 0x80000005 demonstrates that is sended as 5, and there is no error `PROTOCOL_ERROR`:
```
Hypertext Transfer Protocol Version 2 (HTTP/2)
  5. Streams and Multiplexing
    5.1. Stream States
      5.1.1. Stream Identifiers
             source address: 127.0.0.1:45578
             [send] SETTINGS Frame (length:6, flags:0x00, stream_id:0)
             [recv] SETTINGS Frame (length:42, flags:0x00, stream_id:0)
             [send] SETTINGS Frame (length:0, flags:0x01, stream_id:0)
             [recv] SETTINGS Frame (length:0, flags:0x01, stream_id:0)
             [send] HEADERS Frame (length:15, flags:0x05, stream_id:5)
             [recv] HEADERS Frame (length:59, flags:0x04, stream_id:5)
             [recv] DATA Frame (length:6, flags:0x00, stream_id:5)
             [recv] DATA Frame (length:0, flags:0x01, stream_id:5)
             [recv] Timeout
        × 3: Sends stream identifier that takes more than 31 bits
          -> The endpoint MUST response with a connection error of type PROTOCOL_ERROR.
             Expected: GOAWAY Frame (Error Code: PROTOCOL_ERROR)
                       Connection closed
               Actual: DATA Frame (length:0, flags:0x01, stream_id:5)
```

Can you elaborate why the behaviour is different for `0x80000000` and `0x80000005` values?

Reply:
It looks like we effectively cannot perform such a test with `h2spec`. [func validStreamID(streamID uint32) bool](https://github.com/golang/net/blob/master/http2/frame.go#L642) checks if 31-th bit is set, but [framer.AllowIllegalWrites](https://github.com/tempesta-tech/h2spec/blob/master/spec/connection.go#L90) is `true`. Even if we add, say, a flag that would allow us to set it to `false`, that would only cause [func (f *Framer) WriteHeaders(p HeadersFrameParam) error](https://github.com/golang/net/blob/master/http2/frame.go#L1088) to return [errStreamID](https://github.com/golang/net/blob/master/http2/frame.go#L632) error and still no frames would be sent. We can't change this behavior within `h2spec` code.